lenskit-core/src/main/java/org/grouplens/lenskit/data/event/Rating.java

```diff
@@ -20,23 +20,95 @@
  */
 package org.grouplens.lenskit.data.event;
 
-import javax.annotation.Nullable;
-
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
 import org.grouplens.lenskit.data.pref.Preference;
 
+import javax.annotation.Nullable;
+import java.io.Serializable;
+
 /**
- * A rating is an expression of preference for an item by a user.
- *
- * <p>Ratings are equal if they have the same user, item, preference and timestamp.  The hash code
- * of a rating is the hash code of its preference (if it has one), or the user and item ID.
- * Timestamp is <em>ignored</em> in the hash code.  See {@link Ratings#equals(Rating, Rating)} and
- * {@link Ratings#hashRating(Rating)}.</p>
+ * A rating is an expression of preference for an item by a user.  The preference can be {@code null}, in which case
```

Comment:
Why do we indirect through a preference object? From what I can see its only to give us a nullable way to code rating vs. unrating?

Also the separation of rating and unrating makes this class hard to understand and more complex that it is clear to me that we need.